rtl/src/main/scala/tensil/util/decoupled/WidthConverter.scala

```diff
@@ -0,0 +1,107 @@
+package tensil.util.decoupled
+
+import chisel3._
+import chisel3.util.Decoupled
+import chisel3.util.log2Ceil
+import tensil.util
+import chisel3.util.Cat
+
+class WidthConverter(inWidth: Int, outWidth: Int) extends Module {
+  val io = IO(new Bundle {
+    val in  = Flipped(Decoupled(UInt(inWidth.W)))
+    val out = Decoupled(UInt(outWidth.W))
+  })
+
+  if (outWidth == inWidth) {
+    io.out <> io.in
+  } else {
+    val gcd             = util.greatestCommonDivisor(inWidth, outWidth)
+    val lcm             = util.leastCommonMultiple(inWidth, outWidth)
+    val numBlocks       = lcm / gcd
+    val blocksPerInput  = inWidth / gcd
+    val blocksPerOutput = outWidth / gcd
+
+    val arr        = RegInit(VecInit(Array.fill(numBlocks)(0.U(gcd.W))))
+    val enqPtr     = RegInit(0.U(log2Ceil(numBlocks).W))
+    val deqPtr     = RegInit(0.U(log2Ceil(numBlocks).W))
+    val maybeFull  = RegInit(false.B)
+    val doEnq      = WireDefault(io.in.fire)
+    val doDeq      = WireDefault(io.out.fire)
+    val ptrMatch   = enqPtr === deqPtr
+    val enqPtrNext = enqPtr +& blocksPerInput.U
+    val deqPtrNext = deqPtr +& blocksPerOutput.U
+    val full =
+      chisel3.Mux(
+        ptrMatch,
+        maybeFull,
+        chisel3.Mux(
+          enqPtr < deqPtr,
+          enqPtrNext > deqPtr,
+          chisel3.Mux(
+            enqPtrNext > numBlocks.U,
+            (enqPtrNext % numBlocks.U) > deqPtr,
+            false.B
+          )
+        )
+      )
+    val empty = chisel3.Mux(
+      ptrMatch,
+      !maybeFull,
+      chisel3.Mux(
+        deqPtr < enqPtr,
+        deqPtrNext > enqPtr,
+        chisel3.Mux(
+          deqPtrNext > numBlocks.U,
+          (deqPtrNext % numBlocks.U) > enqPtr,
+          false.B
+        )
+      )
+    )
+
+    when(doEnq =/= doDeq) {
+      maybeFull := doEnq
+    }
+
+    when(doEnq) {
+      for (i <- 0 until blocksPerInput) {
+        arr(enqPtr + i.U) := io.in.bits(
+          (i + 1) * gcd - 1,
+          i * gcd
+        )
+      }
+      count(enqPtr, blocksPerInput, numBlocks)
+    }
+
+    when(doDeq) {
+      count(deqPtr, blocksPerOutput, numBlocks)
+    }
+
+    io.in.ready := !full
+    io.out.valid := !empty
+
+    io.out.bits := Cat(
+      (for (i <- 0 until blocksPerOutput)
+        yield arr(deqPtr + i.U)).reverse
+    )
+
+    // TODO handle pipe and flow for max throughput
```

Comment:
Two questions here.

What is the improvement we can see with pipe and flow handled?

Why code below is commented out?

Reply:
I'll just delete it - these ideas come from the chisel queue implementation which I essentially copied and modified to implement this width converter. The advantage is that you could have the output be dequeued on the same cycle as the input, and the input could become ready on the same cycle the output becomes ready. The problem is that it's actually quite complicated to implement in the case of the width converter. It also doesn't really matter as long as we have queues on one or both sides of the converter.